pingcap/tiflow
# tls(dm): support init tls without client kay/certs

### What problem does this PR solve?


Issue Number:  close #5391 #5335

### What is changed and how it works?

* update tidb version  and ajust new API of  FieldType see more in https://github.com/pingcap/tidb/pull/33569
* use tidb/util to init tls connection

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)

1. start a source with tls and enable relay
```yaml
source-id: mysql-replica-01-tls
flavor: ""
enable-gtid: true
enable-relay: true
relay-binlog-name: ""
relay-binlog-gtid: ""
from:
  host: localhost
  user: dm_tls_test
  password: 123456
  port: 3306
  security:
    ssl-ca: "/Users/ehco/Library/Application\ Support/com.tinyapp.DBngin/Engines/mysql/5E575342-71D5-409E-9080-E72881FD0008/ca.pem"
checker:
  check-enable: false
  backoff-rollback: 5m
  backoff-max: 5m

purge:
  interval: 10
  expires: 1
  remain-space: 15

```
2. start relay for this source

```
❯ dmctl query-status -s mysql-replica-01-tls
{
    "result": true,
    "msg": "",
    "sources": [
        {
            "result": true,
            "msg": "no sub task started",
            "sourceStatus": {
                "source": "mysql-replica-01-tls",
                "worker": "worker1",
                "result": null,
                "relayStatus": {
                    "masterBinlog": "(mysql-bin.000001, 10636)",
                    "masterBinlogGtid": "cd3cf208-edc0-11eb-852c-d674c270ea47:1-45",
                    "relaySubDir": "cd3cf208-edc0-11eb-852c-d674c270ea47.000001",
                    "relayBinlog": "(mysql-bin.000001, 10636)",
                    "relayBinlogGtid": "cd3cf208-edc0-11eb-852c-d674c270ea47:1-45",
                    "relayCatchUpMaster": true,
                    "stage": "Running",
                    "result": null
                }
            },
            "subTaskStatus": [
            ]
        }
    ]
```

3 check if the connection from dm-worker is using tls

```
mysql> show processlist;
+------+-------------+-----------------+------+------------------+------+---------------------------------------------------------------+------------------+
| Id   | User        | Host            | db   | Command          | Time | State                                                         | Info             |
+------+-------------+-----------------+------+------------------+------+---------------------------------------------------------------+------------------+
| 1051 | dm_tls_test | localhost:60272 | NULL | Binlog Dump GTID |   80 | Master has sent all binlog to slave; waiting for more updates | NULL             |
| 1055 | root        | localhost:60310 | NULL | Query            |    0 | starting                                                      | show processlist |
+------+-------------+-----------------+------+------------------+------+---------------------------------------------------------------+------------------+
2 rows in set (0.00 sec)



mysql> SELECT * FROM performance_schema.threads WHERE processlist_id = 1051\G
*************************** 1. row ***************************
          THREAD_ID: 1076
               NAME: thread/sql/one_connection
               TYPE: FOREGROUND
     PROCESSLIST_ID: 1051
   PROCESSLIST_USER: dm_tls_test
   PROCESSLIST_HOST: localhost
     PROCESSLIST_DB: NULL
PROCESSLIST_COMMAND: Binlog Dump GTID
   PROCESSLIST_TIME: 156
  PROCESSLIST_STATE: Master has sent all binlog to slave; waiting for more updates
   PROCESSLIST_INFO: NULL
   PARENT_THREAD_ID: NULL
               ROLE: NULL
       INSTRUMENTED: YES
            HISTORY: YES
    CONNECTION_TYPE: SSL/TLS
       THREAD_OS_ID: 12477417
1 row in set (0.01 sec)
```

4. check MySQL tls version
```
mysql> SHOW GLOBAL VARIABLES LIKE 'tls_version';
+---------------+---------------+
| Variable_name | Value         |
+---------------+---------------+
| tls_version   | TLSv1,TLSv1.1 |
+---------------+---------------+
1 row in set (0.00 sec)

```

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update key monitor metrics in both TiCDC document and official document

### Release note 

```release-note
 `fix a bug that can't establish a tls connection when only ssl-ca is set`.
```
